### PR TITLE
Add actionEmitter to handle onbeforeunload

### PR DESCRIPTION
**Pull request by katrynmc**
@AdamEdgett @barnabyc

### CODE REVIEW

File: spec/navigator_spec.js [hunk 1]
```diff
@@ -656,10 +656,10 @@ describe('Navigator', function () {
...
       subject._invokeActions(request, options);
 
-      expect( target.actionOne ).toHaveBeenCalledWith(request, options);
-      expect( target.actionTwo ).toHaveBeenCalledWith(request, options);
-      expect( target.actionThree ).toHaveBeenCalledWith(request, options);
-      expect( target.actionFour ).toHaveBeenCalledWith(request, options);
+      expect( target.actionOne ).toHaveBeenCalledWith(request, options, jasmine.any(Object));
+      expect( target.actionTwo ).toHaveBeenCalledWith(request, options, jasmine.any(Object));
+      expect( target.actionThree ).toHaveBeenCalledWith(request, options, jasmine.any(Object));
+      expect( target.actionFour ).toHaveBeenCalledWith(request, options, jasmine.any(Object));
```
Comment (barnabyc) [inline spec/navigator_spec.js]: `any(Object)` is the JS equivalent of ¯\_(ツ)_/¯

File: src/navigator.js [hunk 2]
```diff
@@ -211,6 +219,19 @@ Navigator.prototype = {
     this.navigate(uri);
   },
 
+  getExitMessage: function () {
+    return compact(map(this._emitters, function (emitter) {
+      return emitter.getExitMessage();
+    })).join(', ') || undefined;
+  },
```
Comment (barnabyc) [inline src/navigator.js]: I'm slightly concerned about the `join(', ')` call resulting in a string that could be less operable than an array of values. Although I may be unclear on the underlying assumptions of the system?

File: spec/navigator_spec.js (hunk 1 above)
Comment (katrynmc) [inline spec/navigator_spec.js]: @barnabyc .....@hojberg made me do it 😩

File: src/navigator.js (hunk 2 above)
Comment (barnabyc) [inline src/navigator.js]: Spoke with @katrynmc in person. Am comfortable with the rationale.

**Comment by barnabyc** [comment]
👍

File: spec/navigator_spec.js (hunk 1 above)
Comment (barnabyc) [inline spec/navigator_spec.js]: Ha, no worries, I appreciate why it's there. Only alternative would be a custom constructor instance  or something but that's overkill.

**Pull request merged**